zarr/util.py

```diff
@@ -532,3 +534,78 @@ def __exit__(self, *args):
 
 
 nolock = NoLock()
+
+
+class PartialReadBuffer:
+    def __init__(self, store_key, chunk_store):
+        self.chunk_store = chunk_store
+        # is it fsstore or an actual fsspec map object
+        assert hasattr(self, chunk_store, "map")
+        self.map = self.chunk_store.map
+        self.fs = self.chunk_store.fs
+        self.store_key = store_key
+        self.key_path = self.map._key_to_str(store_key)
+        self.buff = None
+        self.nblocks = None
+        self.start_points = None
+        self.n_per_block = None
+        self.start_points_max = None
+        self.read_blocks = set()
+
+    def prepare_chunk(self):
+        assert self.buff is None
+        header = self.fs.read_block(self.key_path, 0, 16)
+        nbytes, self.cbytes, blocksize = cbuffer_sizes(header)
+        typesize, _shuffle, _memcpyd = cbuffer_metainfo(header)
+        self.buff = mmap.mmap(-1, self.cbytes)
+        self.buff[0:16] = header
+        self.nblocks = nbytes / blocksize
+        self.nblocks = (
+            int(self.nblocks)
+            if self.nblocks == int(self.nblocks)
+            else int(self.nblocks + 1)
+        )
+        if self.nblocks == 1:
+            self.buff = self.read_full()
+            return
+        start_points_buffer = self.fs.read_block(
+            self.key_path, 16, int(self.nblocks * 4)
+        )
+        self.start_points = np.frombuffer(
+            start_points_buffer, count=self.nblocks, dtype=np.int32
+        )
+        self.start_points_max = self.start_points.max()
+        self.buff[16 : (16 + (self.nblocks * 4))] = start_points_buffer
+        self.n_per_block = blocksize / typesize
+
+    def read_part(self, start, nitems):
+        assert self.buff is not None
+        if self.nblocks == 1:
+            return
+        blocks_to_decompress = nitems / self.n_per_block
+        blocks_to_decompress = (
+            blocks_to_decompress
+            if blocks_to_decompress == int(blocks_to_decompress)
+            else int(blocks_to_decompress + 1)
+        )
+        start_block = int(start / self.n_per_block)
+        wanted_decompressed = 0
+        while wanted_decompressed < nitems:
+            if start_block not in self.read_blocks:
+                start_byte = self.start_points[start_block]
+                if start_byte == self.start_points_max:
+                    stop_byte = self.cbytes
+                else:
+                    stop_byte = self.start_points[self.start_points > start_byte].min()
+                length = stop_byte - start_byte
+                data_buff = self.fs.read_block(self.key_path, start_byte, length)
+                self.buff[start_byte:stop_byte] = data_buff
+                self.read_blocks.add(start_block)
+            if wanted_decompressed == 0:
+                wanted_decompressed += ((start_block + 1) * self.n_per_block) - start
+            else:
+                wanted_decompressed += self.n_per_block
```

Comment:
neither is one of these two.

Reply:
I added this logic for errors that are raised in the case that you have nitems that are less than the size of a block, but span multiple blocks. The `if` clause is encountered on the first block you decompress. It's setting `wanted_decompressed` to the number of items from that first block that are part of nitems, otherwise on following iterations the wanted items for decompression that are read are just set to the number of items in a block since the start point of that block is the first item

For example:
```
z = zarr.zeros((100), chunks=(10))
z[8:12]
```
would failing before I added this logic because the chunk size is 10, and so it would only read the first block and then end because the number read was greater then `nitems` even though the actual `nitems` wanted weren't all read yet. So in this case, line 604 would evaluate to True on the first block and set `wanted_decompressed = 2` which would then allow `while wanted_decompressed<nitems` in line 591 to evaluate to `True` on the next iteration of the loop so it can read then next block to get the 2 items for `nitems` from the following block as well.

Does that make sense?

Reply:
I just added tests to hit this and the above case. I also realized that the example case above won't cause the failure because the chunk isn't big enough for Blosc to split into different compression blocks. The test case I added works for it though

Reply:
+1